[collect] update about.md (#1231)
exercism/rust: concepts/collect/about.md

# About

To understand the `collect` function it's important to know that it is used to resolve an iterator. Recall that an iterator is a sequence of items
that can be chained with other iterators. Iterators are _lazy_, which means they do not execute until they are resolved by a

> [consuming adaptor][consuming adaptor]. Consuming adaptors process all items in the iterator to produce some kind of output.

Reviewer: > they do not execute until they are resolved by a consuming adaptor

I think this is misleading. Consuming adapters are one way to resolve an iterator, but there are others, such as a `for` loop:

```rust
for ch in "hello world".chars() {
  print!("{}", ch);
}
```

Author: Yes, that's a good point. I'll finesse it.

transformation of the collection's items into another iterator. Common iterator adaptors include: [filter][filter], which retains only items from
a collection that satisfy a condition; [map][map], which transforms the type of the items into another type; and [rev][rev], which reverses the
order of the items. Since iterator adaptors yield another iterator, they can be chained together. For instance, to reverse the characters in a
string one might want an iterator of characters and then a reverse iterator, like so

```rust
let rev_chars = "Hello, World!".chars().rev();
```

But `rev_chars` does not give us a `String` with the characters of "Hello, World!" reversed. It gives us an `impl Iterator<Item = char>`. It's a
type which is _prepared_ to yield a sequence of characters, but it doesn't know how to finally assemble them. In this case it doesn't know if we
want a `Vec<char>` or a `String`.

Even though iterator adaptors do so much, they are _lazy_. But this only means that they are like an arrow drawn in a bow. They are not released
until they have a target, called a [consuming adaptor][consuming adaptor], which uses up either a single iterator or the last iterator in a chain

> of iterators.

Reviewer: While I have _several_ notes here, fundamentally all this information should just live in the `Iterator` documentation. Iterators are absolutely a prerequisite concept for collection, and we shouldn't re-teach the concept here.

Suggest replacing all this with something like:

> Recall that iterators are lazy sequences of items. [Consuming adapters][consuming adaptor] process all items in the iterator to produce some kind of output.

Also, adapter is spelled with an e, not an o.

Author: Thank you for your time and attention to this.

Okay to swapping the `Iterator` coverage with your phrase and to cover `FromIterator`.

If you google "adaptor vs adapter" you'll find various opinions on the right spelling. I use "or" for consistency with how it is spelled in TRPL, which I link to, so I opt to keep it.

https://doc.rust-lang.org/book/ch13-02-iterators.html#methods-that-consume-the-iterator
https://doc.rust-lang.org/book/ch13-02-iterators.html#methods-that-produce-other-iterators

> [`collect`][collect] is one such consuming adaptor, and is what can be used to tell `rev_chars` what type of collection we want to end up with. If
> we want a `Vec<char>` we could insert the type into our call to `collect` like so

Reviewer: Let's be particular about what `collect` is.

```suggestion
[`collect`][collect] a consuming adapter which can be used to produce an instance of any type implementing the [`FromIterator` trait](https://doc.rust-lang.org/std/iter/trait.FromIterator.html). That trait is implemented for all the collections in the standard library. 

Because `FromIterator` is implemented for many types, you need to let Rust know what sort of collection you desire. 
If we want a `Vec<char>` we could insert the type into our call to `collect` like so:
```

Author: Okay, I'll fiddle with that.

```rust
    let rev_chars = "Hello, World!".chars().rev().collect::<Vec<char>>();

> // or, since the collection can only be made of chars if it's a Vec, we don't have to specify them and just use
> // let rev_chars = "Hello, World!".chars().rev().collect::<Vec<_>>();

Reviewer: ```suggestion
// since the item type of the `String::chars()` iterator can be inferred to be `char`, we can omit that information
// let rev_chars = "Hello, World!".chars().rev().collect::<Vec<_>>();
```

```

> and `rev_chars` will now be a `Vec<char>`, but the double colons and outer angle brackets (also known as the `turbofish`) which passes the type
> information to the `collect` function is noisier syntax than we need. By specifying the type on the binding itself the `collect` can infer what

Reviewer: ```suggestion
The double colons and outer angle brackets (i.e. `::<_>`) are known as the `turbofish`.
In this case the turbofish is noisier syntax than we need. By specifying the type on the binding itself the `collect` can infer what
```

> collection it should create without the syntactical noise.

Reviewer: ```suggestion
collection it should create with less syntactical noise.
```

```rust
    let rev_chars: Vec<char> = "Hello, World!".chars().rev().collect();
// or
// let rev_chars: Vec<_> = "Hello, World!".chars().rev().collect();
```

If we wanted a `String` instead we could do it like so

```rust
    let rev_chars: String = "Hello, World!".chars().rev().collect();
```

> But what if we want to return a collection from a function instead of letting a binding? We could do

Reviewer: ```suggestion
Rust's type inference is powerful enough that sometimes we don't need an explicit inline note:
```

```rust
fn get_reversed_chars(phrase: &str) -> Vec<char> {
    phrase.chars().rev().collect()
}
```

The return type of the `get_reversed_chars` function lets the `collect` function infer what collection to return. Here, though, we must specify the
`Vec<char>` and not `Vec<_>` or we get the compilation error

```rust
// the type placeholder `_` is not allowed within types on item signatures
// not allowed in type signatures
// replace with the correct return type

> ```

Reviewer: Not sure this section is essential. Students will discover it for themselves if they encounter the situation, and they'll probably have encountered the rule that Rust never performs type inference in function signatures anyway.

Author: I make no assumption that they've encountered the rule, and I'd rather give a proactive heads up than have them stumble onto it. For pedagogical purposes I think it does no harm and may be helpful.

Reviewer: In that case, don't just show the error they'll get, explain why: Rust doesn't perform type inference for function signatures, not because it's an impossible technical challenge, but because this ensures that someone reading the code can know at a glance what types a function consumes and returns.

Although you can sometimes use `collect` without specifying the type, at other times the compiler will state it requires the type for a `collect`, so it

> can be a good habit to be aware of how to specify what type you want to collect.

Reviewer: Also not sure this is essential. Rustc will complain if it can't infer something, and students' natural reluctance to add a bunch of syntax will encourage them to discover the minimal limits of what rustc will accept.

Author: I agree. I basically put it in as a kind of closing thought to keep the ending from seeming abrupt, but I don't feel strongly about it.
